diff --git a/lmarshal.c b/lmarshal.c
index 232eec6..3acbad7 100644
--- a/lmarshal.c
+++ b/lmarshal.c
@@ -378,7 +378,8 @@ static void mar_decode_value
             dec_buf.size = l;
             dec_buf.head = l;
             dec_buf.seek = 0;
-            lua_load(L, (lua_Reader)buf_read, &dec_buf, "=marshal");
+            if (lua_load(L, (lua_Reader)buf_read, &dec_buf, "=marshal") != 0)
+                luaL_error(L, "cannot decode persisted Lua function: %s", lua_tostring(L, -1));
             mar_incr_ptr(l);
 
             lua_pushvalue(L, -1);
